tracer/build/_build/CriticalPathAnalysis/CriticalPathAnalyzer.cs

```diff
@@ -82,11 +118,24 @@ static List<PipelineStage> GetStages(AbsolutePath rootDirectory, PipelineDefinit
         // the successors (stages that depend on us)
         foreach (var currentStage in stages)
         {
-            var stageDefinition = pipeline.Stages.First(x => x.Stage == currentStage.Id);
+            var stageDefinition = pipeline.Stages.FirstOrDefault(x => x.Stage == currentStage.Id);
+            if (stageDefinition is null)
+            {
+                throw new Exception("Could not find stage called " + currentStage.Id);
+            }
 
             foreach (var dependency in stageDefinition.DependsOn)
             {
-                var predecessor = stages.First(x => x.Id == dependency);
+                if (IgnoreStages.Contains(dependency))
+                {
+                    continue;
+                }
+
+                var predecessor = stages.FirstOrDefault(x => x.Id == dependency);
+                if (predecessor is null)
+                {
+                    throw new Exception(stageDefinition.Stage + " depends on stage '" + dependency + "' that could not be found in output results");
```

Comment:
It's not that I think it's a big deal 😁, but is there any opposition to using string interpolation here?
```suggestion
                    throw new Exception($"{stageDefinition.Stage} depends on stage '{dependency}' that could not be found in output results");
```

would make it a lot more readable/maintenable in my opinion.